adr/0002-masking-configuration-API.md

```diff
@@ -0,0 +1,199 @@
+
+### Allow custom masking configurations per target key
+
+We want to provide ability to customize masking per the target key(s) with the options above and also accommodate for
+custom strategies in the future (i.e. leave last 4 digits, preserve special symbols, mask email with preserving domain, etc.).
+
+### Change the JSON masking configuration API
+
+The API for configuring `json-masker` will change to make it more consistent and extendable and allow customizations
+per JSON type.
+
+We will provide configuration options to customize all the defaults as well as the ability to restore the original
+behaviour of pre `1.0.0`.
+
+## Consequences
+
+Therefore the API will be changed to allow these masking options:
+```java
+public class MaskingConfig {
+    // Number masking, mutually exclusive options
+    private Boolean disableNumberMasking = false;    // disables number masking
+    private String maskNumbersWithString = "###";    // default
+    private Integer maskNumberWith;                  // mask number with another number, i.e. 0. Preserves the type 
+    private Integer maskNumberDigitsWith;            // mask all digits of a number with masking digit, must be 1 <= x <= 9 to avoid leading zeroes, preserves original length
+
+    // String masking, mutually exclusive options
+    private String maskStringsWith = "***";         // default
+    private Char maskStringCharactersWith;          // mask all characters with masking character, preserves original length
+
+    // Boolean masking, mutually exclusive options
+    private Boolean disableBooleanMasking = false;  // disables boolean masking
+    private String maskBooleansWithString = "&&&";  // default
+    private Boolean maskBooleanWith;                // mask boolean with another boolean, i.e. false. Preserves the type
+    
+    public static class Builder {
+        ...
+        public Builder disableNumberMasking() { ... };
+        public Builder maskNumbersWith(String) { ... };
+        public Builder maskNumbersWith(int) { ... };
+        public Builder maskNumberDigitsWith(int) { ... };
+
+        public Builder maskStringsWith(String) { ... };
+        public Builder maskStringCharactersWith(char) { ... };
+
+        public Builder disableBooleanMasking() { ... };
+        public Builder maskBooleanWith(String) { ... };
+        public Builder maskBooleanWith(boolean) { ... };
+    }
+} 
+```
+
+Current [JsonMaskingConfig](../src/main/java/dev/blaauwendraad/masker/json/config/JsonMaskingConfig.java) will be changed 
+to have a global per type configuration as well as per-key configuration:
+```java
+public class JsonmaskingConfig {
+    private final Set<String> targetKeys;
+    private final TargetKeyMode targetKeyMode;
+    private final MaskingConfig defaultConfig;
+    private final Map<String, MaskingConfig> targetKeyConfigs;
+    
+    public static class Builder {
+        ...
+
+        // mask mode
+        public Builder mask(String... targets) { ... };
```

Comment:
Should we make the type of keys collection consistent with other `mask` overloads? That is, either use `Set` or `array` everywhere.

Reply:
@Breus any thoughts?